docs/SAFE-SULONG.md

```diff
@@ -0,0 +1,218 @@
+
+   seed = (seed * IA + IC) % IM;
+   return max * seed / IM;
+}
+
+static char nextc()
+{
+   float r;
+   struct aminoacid *iter;
+
+   r = rng(1.0f);
+   iter = lu[(int)(r * SLOTS)];
+   while (iter->p < r)
+      ++iter;
+   return iter->c;
+}
+
+static void random_fasta(struct aminoacid *table, size_t n)
+{
+   size_t i, lines = n / LINELEN;
+   const size_t chars_left = n % LINELEN;
+   char buf[LINELEN + 1];
+
+   while (lines--) {
+      for (i = 0; i < LINELEN; ++i) {
+         buf[i] = nextc();
+      }
+      buf[i] = '\n';
+      fwrite_unlocked(buf, i + 1, 1, stdout);
+   }
+   for (i = 0; i < chars_left; ++i)
+      buf[i] = nextc();
+   fwrite_unlocked(buf, i, 1, stdout);
+}
+
+int main(int argc, char **argv)
+{
+   const size_t n = (argc > 1) ? atoi(argv[1]) : 1000;
+   const char alu[] =
+      "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTG"
+      "GGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGA"
+      "GACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAA"
+      "AATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAAT"
+      "CCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAAC"
+      "CCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTG"
+      "CACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";
+   struct aminoacid iub[] = {
+      { 'a', 0.27 },
+      { 'c', 0.12 },
+      { 'g', 0.12 },
+      { 't', 0.27 },
+      { 'B', 0.02 },
+      { 'D', 0.02 },
+      { 'H', 0.02 },
+      { 'K', 0.02 },
+      { 'M', 0.02 },
+      { 'N', 0.02 },
+      { 'R', 0.02 },
+      { 'S', 0.02 },
+      { 'V', 0.02 },
+      { 'W', 0.02 },
+      { 'Y', 0.02 },
+      { 0, 0 }
+   };
+   struct aminoacid homosapiens[] = {
+      { 'a', 0.3029549426680 },
+      { 'c', 0.1979883004921 },
+      { 'g', 0.1975473066391 },
+      { 't', 0.3015094502008 },
+      { 0, 0 }
+   };
+
+   fputs_unlocked(">ONE Homo sapiens alu\n", stdout);
+   repeat_fasta(alu, n * 2);
+   fputs_unlocked(">TWO IUB ambiguity codes\n", stdout);
+   acc_probs(iub);
+   random_fasta(iub, n * 3);
+   fputs_unlocked(">THREE Homo sapiens frequency\n", stdout);
+   acc_probs(homosapiens);
+   random_fasta(homosapiens, n * 5);
+   putchar_unlocked('\n');
+
+   return 0;
+}
+```
+
+In the `acc_probs` function the first loop computes accumulates probabilities:
+
+```
+   while ((++iter)->c) {
+      iter->p += (iter-1)->p;
+   }
+```
+
+The second last element should have an accumulated probability of `1`.
+For example, when `iub` is passed to the function (with 16 elements),
+then the function expects the last element `(table + 15)->p` to be `1`.
+Due to rounding errors, the last element is actually `0.999999762`.
+
+This rounding errors manifests itself in a spatial memory safety violation
+in a second loop:
+
+```
+   for (int i = 0; i <= SLOTS; ++i) {
+      while (i > (table->p * SLOTS))
+         ++table;
+      lu[i] = table;
+   }
+```
+
+When `i=SLOTS (4095)` the condition should be false, since `4095 > (1 * 4095)`.
+However, since the last element has the probability `0.999999762` the condition
```

Comment:
check for all occurrences of `last` ;)
Or maybe say you don't consider the null real-last element